apis/objectstorage.k8s.io/v1alpha1/types.go

```diff
@@ -0,0 +1,301 @@
+
+type BucketAccessClassList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []BucketAccessClass `json:"items"`
+}
+
+type BucketAccessSpec struct {
+	// +optional
+	BucketInstanceName string `json:"bucketInstanceName,omitempty"`
+	// +optional
+	BucketAccessRequest string `json:"bucketAccessRequest,omitempty"`
+	// +optional
+	ServiceAccount string `json:"serviceAccount,omitempty"`
+	// +optional
+	MintedSecretName string `json:"mintedSecretName,omitempty"`
+
+	PolicyActionsConfigMapData string `json:"policyActionsConfigMapData,omitempty"`
+
+	// +optional
+	Principal string `json:"principal,omitempty"`
+
+	// +optional
+	Provisioner string `json:"provisioner,omitempty"`
+	// +optional
+	Parameters map[string]string `json:"parameters,omitempty"`
+}
+
+type BucketAccessStatus struct {
+	// +optional
+	Message string `json:"message,omitempty"`
+	// +optional
+	AccessGranted bool `json:"accessGranted,omitempty"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Cluster
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+
+type BucketAccess struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec BucketAccessSpec `json:"spec,omitempty"`
+	// +optional
+	Status BucketAccessStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type BucketAccessList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []BucketAccess `json:"items"`
+}
+
+type BucketAccessRequestSpec struct {
+	// +optional
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+
+	BucketRequestName string `json:"bucketRequestName"`
+
+	BucketAccessClassName string `json:"bucketAccessClassName"`
+
+	// +optional
+	BucketAccessName string `json:"bucketAccessName,omitempty"`
+}
+
+type BucketAccessRequestStatus struct {
+	// +optional
+	Message string `json:"message,omitempty"`
+	// +optional
+	AccessGranted bool `json:"accessGranted"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Namespaced
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+
+type BucketAccessRequest struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec BucketAccessRequestSpec `json:"spec,omitempty"`
+	// +optional
+	Status BucketAccessRequestStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type BucketAccessRequestList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []BucketAccessRequest `json:"items"`
+}
+
+type ObjectReference struct {
```

Comment:
Why are we reinventing this?

Reply:
We are defining a smaller footprint than the core.ObjectReference. Currently it is used as common struct to represent BucketRequest in BucketSpec and PolicyActionsConfigMap in BucketAccessClass. These two references can as well be v1alpha1.BucketRequest and core.ConfigMap. 
Please suggest.

Reply:
Makes sense. Can we call it BucketReference

Reply:
fixed, added as per the suggestion